### ISSUE TITLE
Request: Expansion on Save State to SD ability

### ISSUE THREAD

**Issue by 40wattrange** [MiSTer-devel/Main_MiSTer#579]
Request an option/ability in which we could keep the save state data on the SD card as a file with the ability to load it into a state slot at a later time. Essentially like flash carts.

This is not a request to add more than 4 states, but more a way to be able to keep the save state data file whether it is loaded or not, and then the ability to retrieve and load it to one of the 4 slots via the menu.

**Comment by birdybro** [comment]
The savestates are kept stored on your sd card. Check `/media/fat/savestates/`

![image](https://user-images.githubusercontent.com/16388068/162109732-e9e1cea9-6806-4337-b406-237b07a3ac31.png)

**Comment by 40wattrange** [comment]
Yes, that's where I had this idea. If there was a way to keep them there as
new were made, and a way to repair them.

For instance you could have 10 save state files, and choose one to load to
slot 1, slot 3, etc.

**Comment by sorgelig** [comment]
Current handling is optimal for menu system MiSTer has. It's tuned to have intuitive control with minimum settings.

**Comment by 40wattrange** [comment]
Thanks for the responding!

I figured it could remain simple if there were just a simple save/load menu. At that point it wouldnt be anymore complicated than loading a rom or loading a video preset. Doesnt need anything complex like renaming abilities. If someone wanted to go that far with it they could just rename the file on the SD card as needed.

**Comment by thehughhefner** [comment]
I believe this ticket can be closed since @40wattrange asked for being able to save savestates to SD card and loading them later. This is currently possible and you can have multiple savestates that are differentiated by the number after the underscore. Having a secondary menu where you can scroll down and select savestates will be challenging and will be resource-heavy

**Comment by 40wattrange** [comment]
My request wasn't to have multiple save state slots, but more loading a
state to that slot. Right now it's limited to the 4 slots themselves.

**Comment by thehughhefner** [comment]
Confirm if I understand it correctly, you want more than 4 save state slots? If so then there is a reason for it. The MiSTer OSD menu design is limited due to the fact that "pressing more than 4 times" is tedious because you want to go back to the first slot (since it's linear, only goes one way). Unless there is the same functionality added that Zakk implemented with the filters where a user can toggle between filters using the L and R buttons (goes both forward and backward).

**Comment by 40wattrange** [comment]
No this is not what I am asking.

To clarify, we would be fine with 1 save state slot instead of the current
4. We dont want extra slots, we just want the ability to have numerous
states as files to be loadable. Currently if you save the state to the SD
card - it will just be overwritten when you make the next save state in
that slot. What I (and speedrunners) are requesting is the ability to keep
the states on the SD card rather than overwriting them when you use that
slot again. Similar to how we load roms, load audio filters, or load video
filters. Make a save state, it sits on the SD card as "save 1" - make a new
state as "save 02" etc. Now you have 99 states as files on the sd card -
not slots - just the files.

Within the menu you just say "load state" and it pulls from that file and
loads. Just the ability to keep the files on the SD card and load them to
the slot as needed without them being overwritten and limited to 4 files -
again files, not slots.

Right now I can do this manually with WinSCP - just copy the save state
files to my computer, rename them, and when needed I just upload the file
and rename it back to whatever slot name. Again - not needing extra slots -
just keeping the save state file, and pushing it when needed rather than
being replaced on the next save state creation.

Essentially how a modern everdrive handles multiple save states, but only
having one slot to load them.

**Comment by sorgelig** [comment]
When you like to enjoy a bike riding, then you just get a generic bike. If you need bike for sport then you take a special bike for that. MiSTer is not tuned for speed runners and other very specialized usages. Get emulator with billions options and features if you need special use. Purpose of MiSter is recreation of original system as close as possible. What you call "convenience" for speed runners is not exactly convenience for normal use. More options - more chance to get incorrect state, more complicated for generic use, more complicated menu.

**Comment by 40wattrange** [comment]
Sounds good - was just clarifying the ask from @thehughhefner

**Comment by thehughhefner** [comment]
Ok now I get what you mean. Aka save stating but not overwriting the state slot (each save state increases in number 1-99). I still think what is currently available is good but maybe also having the option as you mention. Two things out of this:
- Limited resource for adding this.
- This ticket should probably be added to the respective cores that support save state since save state is not a framework feature. @RobertPeip could probably chime in here.

**Comment by 40wattrange** [comment]
I originally filed this in the NES core repository and was instructed to move it here.

https://github.com/MiSTer-devel/NES_MiSTer/issues/277

**Comment by RobertPeip** [comment]
It IS a framework feature, as the core only knows 4 savestate slots, which are all available as memory area in DDR3. When the header of one slot is changed, then HPS/Framework/mistermain will notice by polling from time to time and save to sd card.

There was a proposal of a more comfortable savestate system from Zakk on discord, where you could have seperate files, but i'm not sure how far he has gotten or if he will complete it.

Overall, having something that works both simple, but also allows for more flexility is not easy to do. Also it should work with all kinds of cores, so just adding some functionality is not useful.

**Comment by zakk4223** [comment]
https://user-images.githubusercontent.com/22642/168107055-b7c968fb-8c76-4880-9d6d-985392322b01.mp4
